### PR TITLE
feat: eventing teskit for java sdk

### PR DESCRIPTION
**Pull request by efgpinto**
Opened this draft PR on the side to track the discussion on what would mean to have no upper type on the `expect` methods.

This would mean that both users from protocol-first or code-first would use the same API. I think this could work but opening for discussion.

~This is draft only, so might be missing some edge cases.~

### CODE REVIEW

File: samples/java-spring-eventsourced-counter/src/it/java/com/example/CounterIntegrationTest.java [hunk 1]
```diff
@@ -52,6 +70,34 @@ public void verifyCounterEventSourcedWiring() {
...
+                .retrieve()
+                .bodyToMono(String.class)
+                .block(timeout);
+        assertEquals("\"200\"", counterMultiply);
+
+        var eventIncreased = outTopic.expectOneTyped(CounterEvent.ValueIncreased.class);
+        assertEquals(new CounterEvent.ValueIncreased(10), eventIncreased.getPayload());
+
+        var eventMultiply = outTopic.expectOneTyped(CounterEvent.ValueMultiplied.class);
+        assertEquals(new CounterEvent.ValueMultiplied(20), eventMultiply.getPayload());
```
Comment (efgpinto) [inline samples/java-spring-eventsourced-counter/src/it/java/com/example/CounterIntegrationTest.java:99]: note this is a java sdk and we are able to use the same exact API

File: sdk/java-sdk-protobuf-testkit/src/main/scala/kalix/javasdk/testkit/impl/EventingTestKitImpl.scala [hunk 2]
```diff
@@ -150,37 +150,48 @@ class TopicImpl(private val probe: TestProbe, codec: MessageCodec) extends Topic
...
+    val contentType = metadata.get("Content-Type").asScala
+
+    (ceType, contentType) match {
+      case (_, Some("text/plain; charset=utf-8")) => "type.kalix.io/string"
+      case (_, Some("application/octet-stream"))  => "type.kalix.io/bytes"
+      case (Some(t), Some("application/json"))    => s"json.kalix.io/$t"
+      case (Some(t), _)                           => s"type.googleapis.com/$t"
+      case (t, ct) =>
+        log.warn(s"Could not extract typeUrl from ce-type=$t content-type=$ct")
+        ""
```
Comment (efgpinto) [inline sdk/java-sdk-protobuf-testkit/src/main/scala/kalix/javasdk/testkit/impl/EventingTestKitImpl.scala:193]: we are probably still missing some edge cases here I think.. I will need to double check and cover with some tests.

File: sdk/java-sdk-protobuf-testkit/src/main/scala/kalix/javasdk/testkit/impl/EventingTestKitImpl.scala [hunk 3]
```diff
@@ -150,37 +150,48 @@ class TopicImpl(private val probe: TestProbe, codec: MessageCodec) extends Topic
...
-  override def expectOneTyped[T <: GeneratedMessageV3](clazz: Class[T], timeout: time.Duration): TestKitMessage[T] = {
+  override def expectOneTyped[T](clazz: Class[T], timeout: time.Duration): TestKitMessage[T] = {
     val msg = probe.expectMsgType[EmitSingleCommand]
     val metadata = new MetadataImpl(msg.getMessage.getMetadata.entries)
-    val decodedMsg = codec.decodeMessage(ScalaPbAny(typeUrlFor(metadata), msg.getMessage.payload))
+    val scalaPb = ScalaPbAny(typeUrlFor(metadata), msg.getMessage.payload)
+
+    val decodedMsg = if (typeUrlFor(metadata).startsWith(JsonSupport.KALIX_JSON)) {
+      val objectMapper = JsonSupport.getObjectMapper
+      objectMapper.readerFor(clazz).readValue(msg.getMessage.payload.toByteArray)
```
Comment (octonato) [inline sdk/java-sdk-protobuf-testkit/src/main/scala/kalix/javasdk/testkit/impl/EventingTestKitImpl.scala:163]: We can use `JsonSupport.decodeJson` directly and it has already some extra logic to deal with primitive bytes

Comment (efgpinto) [reply sdk/java-sdk-protobuf-testkit/src/main/scala/kalix/javasdk/testkit/impl/EventingTestKitImpl.scala:163]: Yeah, was the first approach actually but somehow it fails when deserializing the message.. as if the format was not correct. Haven't found the reason why. Will give it another try.

Comment (octonato) [reply sdk/java-sdk-protobuf-testkit/src/main/scala/kalix/javasdk/testkit/impl/EventingTestKitImpl.scala:275]: Have you tried it again? 
Was that related to the extra characters in front of the payload?

File: sdk/java-sdk-protobuf-testkit/src/main/scala/kalix/javasdk/testkit/impl/EventingTestKitImpl.scala [hunk 4]
```diff
@@ -322,15 +335,23 @@ private[testkit] class TopicImpl(
   override def publish(message: ByteString, metadata: SdkMetadata): Unit =
     brokerProbe.emit(message, metadata)
 
-  override def publish[T <: GeneratedMessageV3](message: TestKitMessage[T]): Unit =
-    publish(message.getPayload.toByteString, message.getMetadata)
+  override def publish(message: TestKitMessage[?]): Unit = message.getPayload match {
```
Comment (octonato) [inline sdk/java-sdk-protobuf-testkit/src/main/scala/kalix/javasdk/testkit/impl/EventingTestKitImpl.scala:338]: I don't remember to have ever seen `[?]` usage in Scala. Is that the same as `[_]`?

File: sdk/java-sdk-protobuf-testkit/src/main/scala/kalix/javasdk/testkit/impl/EventingTestKitImpl.scala [hunk 5]
```diff
@@ -322,15 +335,23 @@ private[testkit] class TopicImpl(
...
+    case javaPb: GeneratedMessageV3 => publish(javaPb.toByteString, message.getMetadata)
+    case scalaPb: GeneratedMessage  => publish(scalaPb.toByteString, message.getMetadata)
+    case str: String                => publish(ByteString.copyFromUtf8(str), message.getMetadata)
+    case _ =>
+      val encodedMsg = JsonSupport.getObjectMapper
+        .writerFor(message.getPayload.getClass)
+        .writeValueAsBytes(message.getPayload)
+      publish(ByteString.copyFrom(encodedMsg), message.getMetadata)
```
Comment (octonato) [inline sdk/java-sdk-protobuf-testkit/src/main/scala/kalix/javasdk/testkit/impl/EventingTestKitImpl.scala:346]: Same comment as above, we should use `JsonSupport.encodeJson` instead. We must understand why this is not working.

**Review by octonato — COMMENTED** [review]
Looking good. 

Good to not depend on `GeneratedMessageV3`

File: sdk/java-sdk-protobuf-testkit/src/main/scala/kalix/javasdk/testkit/impl/EventingTestKitImpl.scala (hunk 4 above)
Comment (efgpinto) [reply sdk/java-sdk-protobuf-testkit/src/main/scala/kalix/javasdk/testkit/impl/EventingTestKitImpl.scala:338]: Me neither, it was a typo but now I learnt that's what we'll actually need to use in scala 3... see [this](https://docs.scala-lang.org/scala3/reference/changed-features/wildcards.html#)

File: sdk/java-sdk-protobuf-testkit/src/main/scala/kalix/javasdk/testkit/impl/EventingTestKitImpl.scala (hunk 3 above)
Comment (efgpinto) [reply sdk/java-sdk-protobuf-testkit/src/main/scala/kalix/javasdk/testkit/impl/EventingTestKitImpl.scala:275]: So, after some experiments, I think it makes sense to not use `JsonSupport.decodeJson` here as that expects the payload to be encoded with the special tag. However, when we publish the messages out to a broker we don't include such tag, and that's why using the `decodeJson` would break here. To make it work, we would need to manually wrap the payload with `AnySupport.encodingBytesPrimitive` in line 271 and then we could use the `decodeJson` but I don't see a point in doing that.

File: sdk/java-sdk-protobuf-testkit/src/main/scala/kalix/javasdk/testkit/impl/EventingTestKitImpl.scala (hunk 5 above)
Comment (efgpinto) [reply sdk/java-sdk-protobuf-testkit/src/main/scala/kalix/javasdk/testkit/impl/EventingTestKitImpl.scala:346]: Similarly to the other occurrence but reversed, here we are offering to serialize the object for the user and in a real scenario where the user is writing to a topic, the payload will not be preceded by the special proto tag that "encodeJson" includes, so I don't think we should use it here at all. Also, note that the user is not forced to use our serialization, it can use `publish(ByteString, Metadata)` and push whatever bytes they want directly.

Comment (johanandren) [reply sdk/java-sdk-protobuf-testkit/src/main/scala/kalix/javasdk/testkit/impl/EventingTestKitImpl.scala:346]: This seams reasonable, given that `publish(Array[Byte])` puts exactly those bytes in the pretend topic, and presents those bytes to eventing in the proxy, just like if they came from a message broker. The proxy will then TLA-encode the bytes for passing around and possibly hand those bytes to an action for example.

Same should go in the other direction, the proxy should decode `Any` with our custom type_url into non-prefixed bytes when publishing to a topic.

There might be something slightly related to attaching metadata with content type `application/json` when auto-transforming like this and just publishing the raw bytes like if to a broker topic.

Comment (johanandren) [reply sdk/java-sdk-protobuf-testkit/src/main/scala/kalix/javasdk/testkit/impl/EventingTestKitImpl.scala:346]: I see that metadata magic is in the TestkitMessageImpl

**Review by octonato — APPROVED** [review]
LGTM
